File: src/Atlasd/Battlenet/Protocols/Game/Messages/SID_CHECKDATAFILE.cs
using Atlasd.Battlenet.Exceptions;
using Atlasd.Daemon;
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace Atlasd.Battlenet.Protocols.Game.Messages
{
    class SID_CHECKDATAFILE : Message
    {
        public enum RequestIds : UInt32
        {
            TermsOfService_usa = 0x01,
            BnServerListW3 = 0x03,
            TermsOfService_USA = 0x1A,
            BnServerList = 0x1B,
            IconsSC = 0x1D,
            BnServerListD2 = 0x80000004,
            ExtraOptionalWorkIX86 = 0x80000005,
            ExtraRequiredWorkIX86 = 0x80000006,
        };

        public SID_CHECKDATAFILE()
        {
            Id = (byte)MessageIds.SID_CHECKDATAFILE;
            Buffer = new byte[0];
        }

        public SID_CHECKDATAFILE(byte[] buffer)
        {
            Id = (byte)MessageIds.SID_CHECKDATAFILE;
            Buffer = buffer;
        }

        public override bool Invoke(MessageContext context)
        {
            switch (context.Direction)
            {
                case MessageDirection.ClientToServer:
                    {
                        Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] SID_CHECKDATAFILE ({4 + Buffer.Length} bytes)");

                        /**
                         * (UINT8) [20] File Checksum (XSHA-1)
                         * (STRING) Filename
                         */

                        if (Buffer.Length < 21)
                            throw new GameProtocolViolationException(context.Client, "SID_CHECKDATAFILE buffer must be at least 21 bytes");

                        using var m = new MemoryStream(Buffer);
                        using var r = new BinaryReader(m);

                        var fileChecksum = r.ReadBytes(20);
                        var fileName = r.ReadByteString();

                        return new SID_CHECKDATAFILE().Invoke(new MessageContext(context.Client, MessageDirection.ServerToClient));
                    }
                case MessageDirection.ServerToClient:
                    {
                        /**
                         * (UINT32) Status:
                         *     0 - Rejected
                         *     1 - Approved
                         *     2 - Ladder Approved
                         */

                        Buffer = new byte[4];

                        using var m = new MemoryStream(Buffer);
                        using var w = new BinaryWriter(m);

                        w.Write((UInt32)0); // Reject everything from this deprecated message.

                        Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] SID_CHECKDATAFILE ({4 + Buffer.Length} bytes)");
                        context.Client.Send(ToByteArray(context.Client.ProtocolType));
                        return true;
                    }
            }

            return false;
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/Messages/SID_CHECKDATAFILE.vb
Imports AtlasV.Battlenet.Exceptions
Imports AtlasV.Daemon
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Text

Namespace AtlasV.Battlenet.Protocols.Game.Messages
    Class SID_CHECKDATAFILE
        Inherits Message

        Public Enum RequestIds As UInt32
            TermsOfService_usa = &H1
            BnServerListW3 = &H3
            rereTermsOfService_USA = &H1A
            BnServerList = &H1B
            IconsSC = &H1D
            BnServerListD2 = &H80000004UI
            ExtraOptionalWorkIX86 = &H80000005UI
            ExtraRequiredWorkIX86 = &H80000006UI
        End Enum

        Public Sub New()
            Id = CByte(MessageIds.SID_CHECKDATAFILE)
            Buffer = Array.Empty(Of Byte)()
        End Sub

        Public Sub New(ByVal varBuffer As Byte())
            Id = CByte(MessageIds.SID_CHECKDATAFILE)
            Buffer = varBuffer
        End Sub

        Public Overrides Function Invoke(ByVal context As MessageContext) As Boolean
            Select Case context.Direction
                Case MessageDirection.ClientToServer
                    Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
                    If Buffer.Length < 21 Then Throw New GameProtocolViolationException(context.Client, $"{MessageName(Id)} buffer must be at least 21 bytes")

                    Using m = New MemoryStream(Buffer) 'Apparently you dont know this, but this does nothing
                        Using r = New BinaryReader(m)
                            Dim fileChecksum = r.ReadBytes(20)
                            Dim fileName = r.ReadByteString()
                        End Using
                    End Using

                    Return New SID_CHECKDATAFILE().Invoke(New MessageContext(context.Client, MessageDirection.ServerToClient))
                Case MessageDirection.ServerToClient
                    Buffer = New Byte(3) {}

                    Using m = New MemoryStream(Buffer)
                        Using w = New BinaryWriter(m)
                            w.Write(CUInt(0))
                        End Using
                    End Using

                    Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
                    context.Client.Send(ToByteArray(context.Client.ProtocolType))
                    Return True
            End Select

            Return False
        End Function
    End Class
End Namespace
